🏗️ Build spider: Tarrant County Commissioners Court

msrezaie opened the pull request:
## What's this PR do?
<!-- eg. This PR updates the scraper for Cleveland City Council because of changes to how they display their meeting schedule. -->
Adds a spider to scrape meeting information of Tarrant County Commissioners Court. The scraper uses two API endpoints to get the data via POST requests.

## Why are we doing this?
<!-- eg. The website's layout was recently updated, causing our existing scraper to fail. This change ensures our scraper remains functional and continues to provide timely updates on council meetings. -->
Scraper requested from [spreadsheet](https://docs.google.com/spreadsheets/d/1-HwPQKwm56vZ8h7CEM65GazKLTS0086gmaEzEHXp7O4/edit?usp=sharing).

## Steps to manually test
1. Ensure the project is installed:
```
pipenv sync --dev
```
2. Activate the virtual env and enter the pipenv shell:
```
pipenv shell
```
3. Run the spider:
```
scrapy crawl fortx_Tarrant_County_Commissioners_Court -O test_output.csv
```
4. Monitor the output and ensure no errors are raised.
5. Inspect `test_output.csv` to ensure the data looks valid.
6. Ensure all tests pass
```
pytest
```

## Are there any smells or added technical debt to note?
No

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new spider for scraping meeting information from the Tarrant County Commissioners Court.
	- Added structured JSON datasets for archived and upcoming meetings, providing detailed meeting attributes.

- **Tests**
	- Implemented a comprehensive suite of unit tests for the new spider, validating data extraction and formatting accuracy. 

- **Chores**
	- Added mock JSON files for testing purposes, containing meeting data for both archived and upcoming meetings.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

msrezaie committed:
spider for Tarrant County Commissioners Court

msrezaie committed:
isort fix

msrezaie committed:
line length fix

msrezaie committed:
black fix

msrezaie committed:
removed unnecessary conditional

msrezaie committed:
fixed failed tests

lamle-ea commented on city_scrapers/spiders/fortx_Tarrant_County_Commissioners_Court.py:
```diff
@@ -0,0 +1,95 @@
+            return dateparse(datetime_str)
+        return None
+
+    def _parse_links(self, item):
```
I noticed that in addition to the agenda ID and video ID, the fetched data also includes the minutes ID. I believe we can construct the agenda URL and minutes URL using `agendaAttachmentId` and `minutesAttachmentId` with the following base URL: `https://tarrant-agendamanagement-public.techsharetx.gov/publicportal/api/meetingattachments/download?id={id}` (e.g. https://tarrant-agendamanagement-public.techsharetx.gov/publicportal/api/meetingattachments/download?id=c170623f-9a45-453c-017b-08dcfda5b8b4). What are your thoughts on this approach?

msrezaie replied:
Thanks for the feedback! I didn’t realize there was a single URL structure that could be used for both the agenda and minutes attachments. That’s a great insight. I'll refactor the `_parse_links` logic.

cruznunez commented on city_scrapers/spiders/fortx_Tarrant_County_Commissioners_Court.py:
```diff
@@ -0,0 +1,95 @@
+            )
+
+    def parse(self, response):
+        data = json.loads(response.text)
```
i think scrapy already comes with the ability to convert the response to json. try

```suggestion
        data = response.json()
```

see [L25 on my PR](https://github.com/cruznunez/city-scrapers-losca/blob/spider-losca_City_Council/city_scrapers/spiders/losca_City_Council.py#L25)

cruznunez commented on tests/test_fortx_Tarrant_County_Commissioners_Court.py:
```diff
@@ -0,0 +1,91 @@
+freezer = freeze_time("2024-11-25")
+freezer.start()
+
+parsed_items = [item for item in spider.parse(test_response)]
```
since there are two urls in the `start_urls` array, maybe it makes sense to parse two separate API files in the tests? i'm not 100% sure about the code but something likes this

```suggestion
upcoming_items = [item for item in spider.parse(upcoming_response)]
archived_items = [item for item in spider.parse(archived_response)]
```

msrezaie replied:
It would be better if there are two separate meeting item files since there are two API URLs used. Thanks for the suggestion.

cruznunez reviewed (changes requested):
Nice man. We could remove a dependency and maybe improve the testing, but besides that, looks good! Well done.

msrezaie committed:
made improvements

- used a universal url for fetching meeting attachments
- separated the meeting items into two files since there are two different API urls used to fetch the meeting data

cruznunez commented on city_scrapers/spiders/fortx_Tarrant_County_Commissioners_Court.py:
```diff
@@ -0,0 +1,109 @@
+import json
```
now that we've switched over to `data = response.json()` we should be able to remove this dependency.

msrezaie replied:
the `json` package is still being used in `scrapy.Request` to dump the dictionary with the `committeeId` as a JSON value in [line 44](https://github.com/msrezaie/city-scrapers-fortx/blob/tarrant-county-commissioners-court-spider/city_scrapers/spiders/fortx_Tarrant_County_Commissioners_Court.py#L44).

cruznunez reviewed (changes requested):
Nice updates! Well done testing the extra file. 

We just need to finish removing the json dependency. I know we want to parse the response as json, but we don't need the json library directly to get it. `response.json()` already comes with scrapy and importing json on top of that is unnecessary.

lamle-ea reviewed (approved):
LGTM
